eclipse-collections-api/src/main/java/org/eclipse/collections/api/RichIterable.java

```diff
@@ -1998,6 +1998,55 @@ default void appendString(Appendable appendable, String separator)
      */
     <V> Multimap<V, T> groupBy(Function<? super T, ? extends V> function);
 
+    /**
+     * For each element of the iterable, the functions are evaluated and the results of these valuations are collected
+     * into a new map of Multimap, where the first and second functions are evaluated to produce the keys to the map and the inner
+     * Multimap respectively, and the original values are added to the same (or similar) species of collection as the
+     * source iterable.
+     * <p>
+     * Example using a Java 8 method reference:
+     * <pre>
+     * Map&lt;String, Multimap&lt;String, Person&gt;&gt; peopleByCityAndLastName =
+     *     people.groupBy(Person::getCity, Person::getLastName);
+     * </pre>
+     * <p>
+     * Example using an anonymous inner class:
+     * <pre>
+     * Map&lt;String, Multimap&lt;String, Person&gt;&gt; peopleByCityAndLastName =
+     *     people.groupBy(new Function&lt;Person, String&gt;()
+     *     {
+     *         public String valueOf(Person person)
+     *         {
+     *             return person.getCity();
+     *         }
+     *     },
+     *     new Function&lt;Person, String&gt;()
+     *     {
+     *         public String valueOf(Person person)
+     *         {
+     *             return person.getLastName();
+     *         }
+     *     }
+     *     );
+     * </pre>
+     *
+     * @since 9.x
+     */
+    default <V, W> MutableMap<V, Multimap<W, T>> groupBy2(Function<? super T, ? extends V> function1, Function<? super T, ? extends W> function2)
+    {
+        return this.<V>groupBy(function1).toMap().collectValues((v, t) -> t.groupBy(function2));
```

Comment:
Maybe the default implementation should just throw?

Reply:
Happy to make the changes- although my original thought was the implementing classes of RichIterable can eseentially get them "for free" (implement groupBy and get the additional levels of grouping).  Would that be of a good reason to keep it?

Reply:
Let me give a little more context about this comment.

I think that the covariant return types of this method should closely follow groupByUniqueKey. I previously mentioned groupBy, but groupByUniqueKey is even closer since it returns a map. With that in mind, RichIterable.groupBy2 should return MapIterable (not MutableMap), MutableCollection.groupBy2 should return MutableMap, and ImmutableCollection.groupBy2 should return ImmutableMap.

This implementation is only appropriate for some RichIterables, specifically the mutable ones, so it should move to MutableCollection. With that in mind, I think it makes sense for the method to remain abstract on RichIterable, or if we want to include this change in a minor release, then it should throw.

This implementation delegates to methods that create new collections, and thus it creates extra garbage and rehashes each T. I think we could follow an approach like groupByUniqueKey (GroupByUniqueKeyProcedure) for efficiency.

The implementation is also not overridden for Bags - we usually put in an optimization so that we call hashCode just once for each T instead of each occurrence of T, and test it in RichIterableTestCase.RichIterable_iterationOrder.

Reply:
Thanks for the clarification and the pointers for the implementation

Reply:
For GroupBy2Procedure, in addition to the Mutable target map, should it also take in a supplier to create suitable type of Multimap?

Reply:
Sorry I don't quite follow this. What will the signature be?

Reply:
Looking back at my own comment, I just realized I should have included more details (apologies!).  I have created a GroupBy2Procedure class under package org.eclipse.collections.impl.block.procedure, for use in AbstractRichIterable (similar to how GroupByUniqueKeyProcedure is used for groupByUniqueKey).  

In GroupBy2Procedure the overridden value method, it also instantiate a new Multimap as it encounters a new first level key:

    @Override
    public void value(T object)
    {
        K1 key1 = this.keyFunction1.valueOf(object);
        K2 key2 = this.keyFunction2.valueOf(object);
        this.map.computeIfAbsent(key1, k -> /* instantiate a multimap */ ).put(key2, object);
    }

I just wonder if the procedure can use a FastListMultimap for all cases, or if it has to also vary depending on the original data structure (e.g. similar to how MutableSet's groupBy returns MutableSetMultimap , MutableSet's multi-level groupBys also return MutableSetMultimap under the hood). 

Let me find a good cut-off point (I'm working on the List-based structures) and push the code as it may be easier for code review.

Reply:
According to me, it should vary depending on covariant overrides. ListMultimap for List, BagMultimap for Bag, SetMultimap for Set.

Reply:
Agree, the runtime types should vary based on the original collection. So set.groupBy2() should return a Map of SetMultimaps. Unfortunately, Java won't allow us to similarly vary the compile time types since generics are invariant.

Reply:
Just a thought.  Could we not take the same approach as we did for partition and introduce a new type?  With a method named groupBy2 and groupBy3, you could return a custom structure which hides the underlying map details and makes the generics flat.

It's more work, and we wouldn't want to to make the structure Mutable or Immutable except for perhaps at the bottom Multimap layer.

I haven't thought much yet about the implications of this... but there might be some benefits having an api that perhaps has get(leveOneKey, levelTwoKey) instead of get(levelOneKey).get(levelTwoKey).  It will be safer and more likely to preserve the structure than exposing a Map at each level, especially if the Map is sparse and Mutable where you can just then null out values arbitrality.

One of the benefits I always teach people about Multimap is that it can deal with sparseness and doesn't let you just put null in for a value for a list.  It preserves its shape, and hides lack of a key showing up as a null.

Reply:
@motlin I was not suggesting that in the method signature, but more so at runtime, if a user casts to a the appropriate type it should not throw a ClassCastException. It will be similar to `chunk` wherein the outer most `RichIterable` is always a `MutableList` but the inner batches are of the appropriate type and casts at runtime work fine.

I like @donraab 's idea but I am not too sure how it will work with the overload which accepts a target multimap. It makes sense for `partition()` to not have the flexibility of providing a target collection. However, for groupBy2, groupBy3, we will need a target variant to give the flexibility of controlling the value Multimap.